astropy/modeling/core.py

```diff
@@ -265,6 +266,76 @@ def _create_inverse_property(cls, members):
         cls._inverse = inverse
         del cls.inverse
 
+    def _create_integral_property(cls, members):
+        """
+        Takes any integral defined on a concrete Model subclass (either
+        as a property or method) and wraps it in the generic
+        getter/setter interface for the integral attribute.
+        """
+
+        integral = members.get('integral')
+        if integral is None or cls.__bases__[0] is object:
+            # The latter clause is the prevent the below code from running on
+            # the Model base class, which implements the default getter and
+            # setter for .integral
+            return
+
+        if isinstance(integral, property):
+            # We allow the @property decorator to be omitted entirely from
+            # the class definition, though its use should be encouraged for
+            # clarity
+            integral = integral.fget
+
+        if not callable(integral):
+            raise ModelDefinitionError('Integral must be callable')
+
+        sig = signature(integral)
+        if len(sig.parameters) > 1:
+            integral = cls._create_integral_subclass(integral, sig)
+
+        if six.PY2 and isinstance(integral, types.MethodType):
+            integral = integral.__func__
+
+        # Store the integral getter internally, then delete the given .integral
+        # attribute so that cls.integral resolves to Model.integral instead
+        cls._integral = integral
+        del cls.integral
+
+    def _create_integral_subclass(cls, func, sig):
+        """
+        For Models that take optional arguments for defining their integral,
+        we create a subclass of _Integral with a ``__call__`` method
+        that supports those additional arguments.
+
+        Takes the function's Signature as an argument since that is already
+        computed in _create_integral_property, so no need to duplicate that
+        effort.
+        """
+
+        def __call__(self, **kwargs):
+            return func(self._model, **kwargs)
+
+        kwargs = []
+        for idx, param in enumerate(sig.parameters.values()):
```

Comment:
Indeed. I copied this from `_create_bounding_box_subclass()`, so I'll change this over there as well.